apps/omg/test/support/conformance/merkle_proof_context.ex

```diff
@@ -0,0 +1,208 @@
+# Copyright 2020 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule Support.Conformance.MerkleProofContext do
+  @moduledoc """
+  A package of data associated with a single proof to assert about. Contains the proof, what it proves, and the under-
+  -lying merkle tree leaves as well
+  """
+  defstruct [:leaves, :root_hash, :leaf, :txindex, :proof]
+
+  alias OMG.Merkle
+
+  use PropCheck
+
+  @doc """
+  A correct context - a proof proves something it should
+  """
+  def correct() do
+    let leaves <- such_that(leaves <- list(binary()), when: length(leaves) > 0) do
+      leaves_length = length(leaves)
```

Comment:
Hm, I'm not sold on this
1/ I think we do want duplicates, essentially, nothing prevents such trees from appearing. However, the suggested generator could be an alternative
2/ also, and more importantly, we do want those various binaries being there - varying length, varying padding with zeros (not really there but...), empty binaries. I'd say those are the cases where something weird happening is the largest. Going with `:binary.encode_unsigned` would limit those cases tested

Reply:
I'm not worried about "boring" binaries in leaves because we use cryptographic hash functions and they have "avalanche" property - minimal changes in inputs flip every of the bits of output with 0.5 probability. By providing diverse inputs were are basically testing correctness of used hash function.

Re duplicates - it's just easier to reason about such tree in context of these tests. And ExitId variant derived from hash of tx included into blocks needs to be unique.

Anyway, those are just things to consider.

Reply:
hm, I hear you, but also consider that we're executing code that is "higher level" than just the hashing function. At this higher level, a hash function might even be not there or be misused (in one of the impls). 

Consider a situation where the leaf is for some reason zero-byte-trimmed before entering the proof validating function or erroneously padded. Or a similar scenario on the root_hash/proof generating end. 

(not necessarily saying those tests in this PR cover for these scenarios that well)

Reply:
As noted elsewhere, I'll add the ideas to expand into the (already existing) bunch of todos of the property tests.

However, for this specifically I think there's a very easy and worth while modification to be made. I'll turn `binary()` generator to `pragmatic_binary()` (name pending) which will be `union([binary, boring_binary])`, which is a bit of a middle-ground of both approaches.